### DISCUSSION TITLE
Starting an application

### DISCUSSION THREAD

**Discussion by Teunderbolt** [rhulme/pico-flashloader#9] [Q&A]
Hello,

I am using your project as an example to write my own bootloader, so I am able to update the pi pico using a network connection. There is only one thing I don't understand yet, and that is the assembler code to start the application.

```
asm volatile 
(
        "mov r0, %[start]\n"
        "ldr r1, =%[vtable]\n"
        "str r0, [r1]\n"
        "ldmia r0, {r0, r1}\n"
        "msr msp, r0\n"
        "bx r1\n"
        :
        : [start] "r" (sStart + 0x100), [vtable] "X" (PPB_BASE + M0PLUS_VTOR_OFFSET)
        :
);

```
I am able to flash a hex file to the pico's flash, but the application won't run at the moment. 
I hope that when you clarify the piece of assembler code I will find the thing that goes wrong in my code.

Kind regards,
Teun

**Comment by rhulme** [comment]
Hi,

That code is effectively the same as used in the `vector_into_flash` function in the SDK - https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/boot_stage2/asminclude/boot2_helpers/exit_from_boot2.S

Here's a commented version:
```
asm volatile (
"mov r0, %[start]\n"    // Get address of application's vector table (start of image + 0x100)
"ldr r1, =%[vtable]\n"  // Get address of RP2040 VTOR register
"str r0, [r1]\n"        // Point RP2040 at application's vector table
"ldmia r0, {r0, r1}\n"  // Put values starting at address r0 into r0 and r1
"msr msp, r0\n"         // Set the stack pointer
"bx r1\n"               // Jump to start of application
:
: [start] "r" (sStart + 0x100), [vtable] "X" (PPB_BASE + M0PLUS_VTOR_OFFSET)
:
);
```
See https://developer.arm.com/documentation/dui0497/a/the-cortex-m0-processor/exception-model/vector-table for the layout of the vector table.  As you can see there, the first two entries are the stack pointer and the entry point.

So it's basically just telling the processor where the application's vector table is and setting up the stack pointer before jumping to the application's entry point.

**Reply by Teunderbolt** [reply]
Hi rhulme,

Thx for your quick reaction, I now fully understand the assembler code, but why are you adding 0x100 to sStart. Does this, have something to do, with the fact that you start writing flash from sStart + offset, were offset has a value of 256? 

I start writing the flash direct from address sStart (offcoarse minus the XIP_BASE). When I run the assembler code with the "+ 0x100", the on-board led will light up but it won't flash. So my thought was, that the start address of the application is wrong.

Have you ever experienced such a problem? If true please help me.

Kind regards,
Teun

**Reply by JaccoVeldscholten** [reply]
I'm really wondering this to. Wrote my own bootloader a few projects ago but I was unable to understand why this problem occured.

**Reply by rhulme** [reply]
Hi,

OK, so I'll go through the boot process from the start.  Maybe you know some of it, maybe not.

When the RP2040 is first powered up, it runs through some code in the on-board ROM.  This configures the processor to be able to speak to external SPI flash (and also implements the "BOOTSEL" routines for updating via USB).  Not all SPI flash chips work exactly the same way.  Some may only support basic SPI, which only has a single data line (per direction), others allow greater throughput using QSPI, which uses four data lines in parallel.  There can be more differences too but I'm only really familiar with them in passing so before I say something that's incorrect, all we need to know is that there are differences between different chips.

The RP2040 cannot know what kind of external flash is connected or what features that chip supports that could improve performance so all it can do is aim for the lowest common denominator that it can rely on all chips supporting.

Remember too that the code in ROM cannot be changed.  The more complicated any piece of code is, the more likely it is that it contains bugs, so you want to keep code in ROM as simple as possible and get out of the ROM as quickly as possible.

Once the ROM bootloader has got basic SPI communication set up with the external flash, it starts running whatever code it finds at the beginning of flash.  This is the second-stage bootloader and is defined to be 256 bytes long.  This second bootloader is part of the image generated when you build an application with the SDK.  The purpose of the second-stage bootloader is to re-configure the SPI communication again and possibly also talk to the external flash to switch it into a different mode that allows better throughput.

When you build your application with the SDK, you can specify what board you are targetting (via the `PICO_BOARD` variable).  By default this will be the Pico but there are other boards out there, or you can define your own.  One of the things that belongs to the board definition is which second-stage bootloader to use (because any given board will always use the same type of flash chip).  There are several different second-stage bootloaders in the SDK.  Look in [rp2_common/boot_stage2](https://github.com/raspberrypi/pico-sdk/tree/master/src/rp2_common/boot_stage2).

The last four bytes of the second-stage bootloader are actually a CRC of the first 252 bytes.  This lets the ROM bootloader detect whether there's a valid image or whether it should fall back to waiting for a new file via USB.  I also use in my flashloader for much the same purpose.

Because the second-stage booloader is in the first 256 (0x100) bytes, that means that everything belonging directly to your application starts after it.  That's why there's the "+ 0x100".

The reason the flashloader starts writing from 256 bytes in is to make the update power-fail safe.  The check for a valid image is just "does the CRC of the first 252 bytes match what's stored in the four bytes immediately afterwards?"  It's not checking the entire image.  If you wrote the data starting from the very beginning, any power failure after the first 256 bytes could not be detected and the processor would start running an image that's only partially there.  By writing everything else first and then going back, you know that if the CRC is OK, the rest of the image must have been successfully flashed.

I'm assuming you're testing the with standard 'blink' demo.  Have you adjusted the build process for that so that its start address is moved?  Things will be linked with absolute addresses starting at 0x10000000 but that's now where the flashloader is so you have to tell the linker to use a different start address.  That's handled by `set_linker_script` in the CMakeLists.txt (look at either the section for app250 or app800).

**Reply by Teunderbolt** [reply]
Hi rhulme,

Again thanks for your quick reaction, and your good explanation. I learned a lot, and debuged a bit more it seems that the pico will controll the pins, but it stalls on a sleep_ms but also on the repeating timer that you are using. Have you ever encounterd something like this? 

I hope to hear form you soon.

Kind regards,
Teun

**Reply by rhulme** [reply]
Hi Teun,

I'm not sure why I asked you whether you were using the standard blink demo.  I assume you're actually using the demo application from my project...

I don't think I've had the problem you're seeing but is it possible you're using the new Pico-W?  The LED on the original board is connected to one of the RP2040's GPIOs, but that GPIO is used for something else on the Pico-W and the LED is controlled via the wireless chip instead.

I just had a quick look at updating the example code to also work with the Pico-W but I think it'll take a little longer as pulling in the required library adds over 300k to the final application size!  I need to see if that can be reduced somehow.

Even if the LED won't currently flash using a Pico-W, the UART interface should still work and you should still see the initial message "Flashing LED every X milliseconds" but maybe it's worth commenting out all the lines that reference `PICO_DEFAULT_LED_PIN` (I don't have a Pico-W to test on).

Otherwise if you say it's stalling, can you post a backtrace?  It should normally be in a busy-loop waiting for data via the UART.

**Reply by Teunderbolt** [reply]
Hi Rhulme,

Thanks for your reply. The board I am using is the regular pi pico from the raspberry orginazation. I haven't found it yet but I will keep debugging until I found the problem. And when I have found it I will let you know offcourse what my mistake was.

Kind regards,
Teun

**Reply by Teunderbolt** [reply]
Hi Rhulme,

During debugging I found out, that the data I readed back from flash always was the same value: 4294967295 in decimal or 0xFFFFFFFF. When I read the data direct from flash using a debugger I got the correct values on the correct addresses, so the rp2040 is writing to flash, but for some reason I can't read from flash. Is there something that needs to be set to enable reading from flash?

Kind regards,
Teun

**Reply by rhulme** [reply]
Hi Teun,

No, you don't need to do anything to enable reading.  Just read from the appropriate address.

You are reading from the correct absolute address, aren't you?  Not from the offset within flash (i.e. the address you pass to `flash_range_erase` and `flash_range_program`)?

Are you seeing this problem with the example code (app.c) or with your own code?  If with your own, have you confirmed that the example works correctly?

Richard

**Reply by Teunderbolt** [reply]
Hi Richard,

As it turns out the debugger is breaking the flash reading function for some reason. The debugger I used is a Segger Jlink EDU V10.1. This is something I didn't expect and this makes debugging a lot harder. I am still researching my problem and if I got something new I will let you know.

Kind regards,
Teun

**Reply by Teunderbolt** [reply]
Hi Richard,

I finaly found my problems, as it turns out I saved and disabled the interrupts in the bootloader this prevented the application to perform a sleep_ms(). You could bypass this by writing the interrupt status to a watchdog scratch register, and recall the scratch register in the application.

Also the use of the stdio.h library prevented the application to run, but this has something todo with my pc (computer, not program counter). When it is initializing the stdio, it is setting up the usb as an serial port, when this is happening my pc is power cycling the usb-bus which sometimes causes a reset.

I hope, that my problom will help other people in the future who are experiencing the same problem.

Kind regards,
Teun

**Reply by rhulme** [reply]
Ah, OK.  You'd changed the bootloader.  Yes, disabling interrupts for any amount of time will often cause problems.
